components/kyma-environment-broker/internal/model.go

```diff
@@ -122,14 +123,31 @@ type Instance struct {
 	Version int
 }
 
+// OperationType defines the possible types of an asynchronous operation to a broker.
+type OperationType string
+
+const (
+	// OperationTypeProvision means provisioning OperationType
+	OperationTypeProvision OperationType = "provision"
+	// OperationTypeDeprovision means deprovision OperationType
+	OperationTypeDeprovision OperationType = "deprovision"
+	// OperationTypeUndefined means undefined OperationType
+	OperationTypeUndefined OperationType = ""
+	// OperationTypeUpgradeKyma means upgrade Kyma OperationType
+	OperationTypeUpgradeKyma OperationType = "upgradeKyma"
+	// OperationTypeUpgradeCluster means upgrade cluster (shoot) OperationType
+	OperationTypeUpgradeCluster OperationType = "upgradeCluster"
+)
+
 type Operation struct {
 	// following fields are serialized to JSON and stored in the storage
 	InstanceDetails
 
-	ID        string    `json:"-"`
-	Version   int       `json:"-"`
-	CreatedAt time.Time `json:"-"`
-	UpdatedAt time.Time `json:"-"`
+	ID        string        `json:"-"`
+	Version   int           `json:"-"`
+	CreatedAt time.Time     `json:"-"`
+	UpdatedAt time.Time     `json:"-"`
+	Type      OperationType `json:"-"`
```

Comment:
why did you add a type to the model? when you have an instance of ProvisioningOperation you know the type is "Provisioning", if you have an instance of DeprovisioningOperation - type is "deprovisioning". where are you using the type?

Reply:
In KEB model, the instance current state is represented in the last operation's current state. There are already some places where we check for concurrent operations to guard against certain conditions - these places actually meant to check the "current instance state".

Now with upgrade kyma and upgrade cluster, there are too much different operation types, and to get the "current instance state", it would be very cumbersome to implement proper conditions for ensuring concurrent operations are not performed in parallel on the same runtime with looking up each (5-6) different operation type specific data from DB and compare. A much natural and easier way is to use the `GetLastOperation()` method, and use the `Operation` object's `Type` and `State` .

I use it mainly [here](https://github.com/kyma-project/control-plane/pull/601/files#diff-53706f0b7a9c8887c58d6376e430c426578722a233dec634591344b110e02cceR70) and [here](https://github.com/kyma-project/control-plane/pull/601/files#diff-69186aec10682e7b7131a19426024d2a79d81cceae25a9b405c2dbdfda0d8e3eR84)